// @delon/abc/src/desc-list/desc-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ElementRef, Renderer2, TemplateRef, ContentChildren, QueryList, } from '@angular/core';
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { DescListItemComponent } from './desc-list-item.component';
import { AdDescListConfig } from './desc-list.config';
var DescListComponent = /** @class */ (function () {
    // endregion
    function DescListComponent(cog, el, renderer) {
        this.el = el;
        this.renderer = renderer;
        // region fields
        this._title = '';
        this._gutter = 32;
        this.layout = 'horizontal';
        this._xs = 24;
        this._sm = 12;
        this._md = 8;
        this._col = 3;
        this._classMap = [];
        Object.assign(this, cog);
    }
    Object.defineProperty(DescListComponent.prototype, "title", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this._title = null;
                this._titleTpl = value;
            }
            else
                this._title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DescListComponent.prototype, "gutter", {
        get: /**
         * 列表项间距，单位为 `px`
         * @return {?}
         */
        function () {
            return this._gutter;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._gutter = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DescListComponent.prototype, "col", {
        get: /**
         * 指定信息最多分几列展示，最终一行几列由 col 配置结合响应式规则决定
         * @return {?}
         */
        function () {
            return this._col;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._col = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DescListComponent.prototype.setClass = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._classMap.forEach(function (cls) {
            return _this.renderer.removeClass(_this.el.nativeElement, cls);
        });
        this._classMap = ['desc-list', this.layout];
        if (this.size)
            this._classMap.push(this.size);
        this._classMap.forEach(function (cls) {
            return _this.renderer.addClass(_this.el.nativeElement, cls);
        });
    };
    /**
     * @return {?}
     */
    DescListComponent.prototype.setResponsive = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ responsive = {
            1: { xs: 24 },
            2: { xs: 24, sm: 12 },
            3: { xs: 24, sm: 12, md: 8 },
            4: { xs: 24, sm: 12, md: 6 },
        }[this.col > 4 ? 4 : this.col];
        this._xs = responsive.xs;
        this._sm = responsive.sm;
        this._md = responsive.md;
    };
    /**
     * @return {?}
     */
    DescListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setClass();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    DescListComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.setClass();
        if (changes.col)
            this.setResponsive();
    };
    DescListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'desc-list',
                    template: "\n  <div *ngIf=\"_title || _titleTpl\" class=\"ad-desc-list__title\">\n    <ng-container *ngIf=\"_title; else _titleTpl\">{{_title}}</ng-container>\n  </div>\n  <div nz-row [nzGutter]=\"gutter\">\n    <div nz-col [nzXs]=\"_xs\" [nzSm]=\"_sm\" [nzMd]=\"_md\" *ngFor=\"let i of _items\">\n      <ng-template [ngTemplateOutlet]=\"i.tpl\"></ng-template>\n    </div>\n  </div>\n  ",
                    host: { '[class.ad-desc-list]': 'true' },
                    preserveWhitespaces: false,
                },] },
    ];
    /** @nocollapse */
    DescListComponent.ctorParameters = function () { return [
        { type: AdDescListConfig, },
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    DescListComponent.propDecorators = {
        "title": [{ type: Input },],
        "size": [{ type: Input },],
        "gutter": [{ type: Input },],
        "layout": [{ type: Input },],
        "col": [{ type: Input },],
        "_items": [{ type: ContentChildren, args: [DescListItemComponent,] },],
    };
    return DescListComponent;
}());
export { DescListComponent };
function DescListComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DescListComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DescListComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    DescListComponent.propDecorators;
    /** @type {?} */
    DescListComponent.prototype._title;
    /** @type {?} */
    DescListComponent.prototype._titleTpl;
    /** @type {?} */
    DescListComponent.prototype.size;
    /** @type {?} */
    DescListComponent.prototype._gutter;
    /** @type {?} */
    DescListComponent.prototype.layout;
    /** @type {?} */
    DescListComponent.prototype._xs;
    /** @type {?} */
    DescListComponent.prototype._sm;
    /** @type {?} */
    DescListComponent.prototype._md;
    /** @type {?} */
    DescListComponent.prototype._col;
    /** @type {?} */
    DescListComponent.prototype._classMap;
    /** @type {?} */
    DescListComponent.prototype._items;
    /** @type {?} */
    DescListComponent.prototype.el;
    /** @type {?} */
    DescListComponent.prototype.renderer;
}
