// @delon/form/src/src/form.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export { DelonFormConfig } from './config';
export { useFactory, SFComponent } from './sf.component';
export { SFItemComponent } from './sf-item.component';
export { SFFixedDirective } from './sf-fixed.directive';
export { DelonFormModule } from './module';
// schemas
export { ERRORSDEFAULT } from './errors';
// widgets
export { FormProperty, PropertyGroup, FormPropertyFactory, AtomicProperty, ObjectProperty, ArrayProperty, StringProperty, NumberProperty, BooleanProperty } from './model/index';
export { Widget, ControlWidget, ArrayLayoutWidget, ObjectLayoutWidget } from './widget';
export { ObjectWidget, ArrayWidget, StringWidget, NumberWidget, DateWidget, TimeWidget, RadioWidget, CheckboxWidget, BooleanWidget, TextareaWidget, SelectWidget, TagWidget, UploadWidget, TransferWidget, SliderWidget, RateWidget, AutoCompleteWidget, CascaderWidget, MentionWidget, CustomWidget, NzWidgetRegistry } from './widgets/index';
export { WidgetRegistry, WidgetFactory } from './widget.factory';
// other
export { SchemaValidatorFactory, AjvSchemaValidatorFactory } from './validator.factory';
