// @delon/form/src/src/sf-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ViewChild, ViewContainerRef, } from '@angular/core';
import { FormProperty } from './model/form.property';
import { WidgetFactory } from './widget.factory';
import { TerminatorService } from './terminator.service';
var /** @type {?} */ nextUniqueId = 0;
var SFItemComponent = /** @class */ (function () {
    // endregion
    function SFItemComponent(widgetFactory, terminator) {
        this.widgetFactory = widgetFactory;
        this.terminator = terminator;
        this.widget = null;
    }
    /**
     * @param {?} widget
     * @return {?}
     */
    SFItemComponent.prototype.onWidgetInstanciated = /**
     * @param {?} widget
     * @return {?}
     */
    function (widget) {
        this.widget = widget;
        var /** @type {?} */ id = "_sf-" + nextUniqueId++;
        var /** @type {?} */ ui = /** @type {?} */ (this.formProperty.ui);
        this.widget.formProperty = this.formProperty;
        this.widget.schema = this.formProperty.schema;
        this.widget.ui = ui;
        this.widget.id = id;
        this.widget.firstVisual = ui.firstVisual;
        this.formProperty.widget = widget;
    };
    /**
     * @return {?}
     */
    SFItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.terminator.onDestroy.subscribe(function () {
            _this.ngOnDestroy();
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    SFItemComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.ref = this.widgetFactory.createWidget(this.container, this.formProperty.ui["widget"] || this.formProperty.schema.type);
        this.onWidgetInstanciated(this.ref.instance);
    };
    /**
     * @return {?}
     */
    SFItemComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.formProperty.ui["__destroy"] = true;
        this.ref.destroy();
    };
    SFItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sf-item',
                    template: "<ng-template #target></ng-template>",
                },] },
    ];
    /** @nocollapse */
    SFItemComponent.ctorParameters = function () { return [
        { type: WidgetFactory, },
        { type: TerminatorService, },
    ]; };
    SFItemComponent.propDecorators = {
        "formProperty": [{ type: Input },],
        "container": [{ type: ViewChild, args: ['target', { read: ViewContainerRef },] },],
    };
    return SFItemComponent;
}());
export { SFItemComponent };
function SFItemComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SFItemComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SFItemComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SFItemComponent.propDecorators;
    /** @type {?} */
    SFItemComponent.prototype.ref;
    /** @type {?} */
    SFItemComponent.prototype.widget;
    /** @type {?} */
    SFItemComponent.prototype.formProperty;
    /** @type {?} */
    SFItemComponent.prototype.container;
    /** @type {?} */
    SFItemComponent.prototype.widgetFactory;
    /** @type {?} */
    SFItemComponent.prototype.terminator;
}
