// @delon/theme/src/theme.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { WINDOW } from './win_tokens';
import { MenuService } from './services/menu/menu.service';
import { ScrollService } from './services/scroll/scroll.service';
import { SettingsService } from './services/settings/settings.service';
import { TitleService } from './services/title/title.service';
import { ALAIN_I18N_TOKEN, AlainI18NServiceFake } from './services/i18n/i18n';
import { _HttpClient } from './services/http/http.client';
var /** @type {?} */ SERVICES = [
    MenuService,
    ScrollService,
    SettingsService,
    TitleService,
    _HttpClient,
];
import { ModalHelper } from './services/modal/modal.helper';
var /** @type {?} */ HELPERS = [ModalHelper];
// components
var /** @type {?} */ COMPONENTS = [];
import { DatePipe } from './pipes/date/date.pipe';
import { CNCurrencyPipe } from './pipes/currency/cn-currency.pipe';
import { KeysPipe } from './pipes/keys/keys.pipe';
import { YNPipe } from './pipes/yn/yn.pipe';
var /** @type {?} */ PIPES = [DatePipe, CNCurrencyPipe, KeysPipe, YNPipe];
// endregion
// region: zorro modules
// import { NzToolTipModule } from 'ng-zorro-antd';
var /** @type {?} */ ZORROMODULES = [];
var AlainThemeModule = /** @class */ (function () {
    function AlainThemeModule() {
    }
    /**
     * @return {?}
     */
    AlainThemeModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AlainThemeModule,
            providers: [
                { provide: WINDOW, useValue: window },
                { provide: ALAIN_I18N_TOKEN, useClass: AlainI18NServiceFake }
            ].concat(SERVICES, HELPERS),
        };
    };
    /**
     * @return {?}
     */
    AlainThemeModule.forChild = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AlainThemeModule,
            providers: HELPERS.slice(),
        };
    };
    AlainThemeModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, RouterModule].concat(ZORROMODULES),
                    declarations: COMPONENTS.concat(PIPES),
                    exports: COMPONENTS.concat(PIPES),
                },] },
    ];
    return AlainThemeModule;
}());
export { AlainThemeModule };
function AlainThemeModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AlainThemeModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AlainThemeModule.ctorParameters;
}
