// @delon/abc/src/zip/interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { InjectionToken } from '@angular/core';
export var /** @type {?} */ DA_ZIP_CONFIG = new InjectionToken('DA_ZIP_CONFIG');
/**
 * @record
 */
export function ZipConfig() { }
function ZipConfig_tsickle_Closure_declarations() {
    /**
     * Zip library path
     * \@default //cdn.bootcss.com/jszip/3.1.5/jszip.min.js
     * @type {?|undefined}
     */
    ZipConfig.prototype.url;
    /**
     * Defines which zip optional utils should get loaded
     * @type {?|undefined}
     */
    ZipConfig.prototype.utils;
}
/**
 * @record
 */
export function ZipWriteOptions() { }
function ZipWriteOptions_tsickle_Closure_declarations() {
    /**
     * save file name, default: `download.zip`
     * @type {?|undefined}
     */
    ZipWriteOptions.prototype.filename;
    /** @type {?|undefined} */
    ZipWriteOptions.prototype.options;
    /**
     * The optional function called on each internal update with the metadata.
     * @type {?|undefined}
     */
    ZipWriteOptions.prototype.update;
    /**
     * triggers when saveas
     * @type {?|undefined}
     */
    ZipWriteOptions.prototype.callback;
}
/**
 * @record
 */
export function ZipSaveOptions() { }
function ZipSaveOptions_tsickle_Closure_declarations() {
    /**
     * 指定保存文件名，默认 `download.zip`
     * @type {?|undefined}
     */
    ZipSaveOptions.prototype.filename;
    /**
     * JSZip `generateAsync` 方法的 `options` 选项
     * @see https://stuk.github.io/jszip/documentation/api_jszip/generate_async.html
     * @type {?|undefined}
     */
    ZipSaveOptions.prototype.options;
    /**
     * JSZip `generateAsync` 方法的 `onUpdate` 回调
     * @see https://stuk.github.io/jszip/documentation/api_jszip/generate_async.html
     * @type {?|undefined}
     */
    ZipSaveOptions.prototype.update;
    /**
     * 保存前回调方法
     * @type {?|undefined}
     */
    ZipSaveOptions.prototype.callback;
}
