// @delon/theme/src/services/settings/interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function App() { }
function App_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    App.prototype.name;
    /** @type {?|undefined} */
    App.prototype.description;
    /** @type {?|undefined} */
    App.prototype.year;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * @record
 */
export function User() { }
function User_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    User.prototype.name;
    /** @type {?|undefined} */
    User.prototype.avatar;
    /** @type {?|undefined} */
    User.prototype.email;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * @record
 */
export function Layout() { }
function Layout_tsickle_Closure_declarations() {
    /**
     * 是否固定顶部菜单
     * @type {?}
     */
    Layout.prototype.fixed;
    /**
     * 是否折叠右边菜单
     * @type {?}
     */
    Layout.prototype.collapsed;
    /**
     * 是否固定宽度
     * @type {?}
     */
    Layout.prototype.boxed;
    /**
     * 语言环境
     * @type {?}
     */
    Layout.prototype.lang;
}
