// @delon/abc/src/lodop/interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function CLodop() { }
function CLodop_tsickle_Closure_declarations() {
    /**
     * 建立打印机名单
     * @type {?}
     */
    CLodop.prototype.Create_Printer_List;
    /**
     * 建立纸张类型名单
     * @type {?}
     */
    CLodop.prototype.Create_PageSize_List;
    /**
     * 判断是否支持https协议的属性
     *
     * - 0 不支持
     * - 1 支持
     * - 2 支持且已启动（https服务需单独启动)
     * @type {?}
     */
    CLodop.prototype.HTTPS_STATUS;
    /**
     * 结果回调函数
     * @type {?}
     */
    CLodop.prototype.On_Return;
    /**
     * 结果回调函数保留
     * @type {?}
     */
    CLodop.prototype.On_Return_Remain;
}
/**
 * @record
 */
export function Lodop() { }
function Lodop_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [key: string]: any;
    */
    /**
     * 获得软件版本号
     * @type {?}
     */
    Lodop.prototype.VERSION;
    /**
     * 打印初始化。初始化运行环境，清理异常打印遗留的系统资源，设定打印任务名。
     *
     * **建议或要求：**该函数与PRINT_INITA都有初始化功能，每个打印事务至少初始化一次，建议打印程序首先调用该函数。任务名要尽量区别于其它打印任务，譬如用“XX单位_XX管理信息系统_XX子系统_XX模块_XX打印作业”字样。
     * 不希望最终用户更改打印布局时，则设strTaskName空。
     *
     * \@param strTaskName 打印任务名
     * \@return 返回逻辑真表示初始化成功，逻辑假表示初始化失败，失败原因有：前一个打印事务没有完成；操作系统没有添加打印机(驱动)等
     * @type {?}
     */
    Lodop.prototype.PRINT_INIT;
    /**
     * 设定纸张大小
     * @type {?}
     */
    Lodop.prototype.SET_PRINT_PAGESIZE;
    /**
     * 增加超文本打印项(普通模式)
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_HTM;
    /**
     * 增加表格打印项（超文本模式）
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_TABLE;
    /**
     * 增加表格打印项（超文本模式）
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_TABLE;
    /**
     * 增加超文本打印项（URL模式）
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_URL;
    /**
     * 增加纯文本打印项
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_TEXT;
    /**
     * 增加图片打印项
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_IMAGE;
    /**
     * 增加矩形线
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_RECT;
    /**
     * 增加椭圆线
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_ELLIPSE;
    /**
     * 增加直线
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_LINE;
    /**
     * 增加条形码
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_BARCODE;
    /**
     * 增加图表
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_CHART;
    /**
     * 装载文档式模板
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_DATA;
    /**
     * 设置打印项风格
     * @type {?}
     */
    Lodop.prototype.SET_PRINT_STYLE;
    /**
     * 打印预览
     * @type {?}
     */
    Lodop.prototype.PREVIEW;
    /**
     * 直接打印
     * @type {?}
     */
    Lodop.prototype.PRINT;
    /**
     * 打印维护
     * @type {?}
     */
    Lodop.prototype.PRINT_SETUP;
    /**
     * 打印设计
     * @type {?}
     */
    Lodop.prototype.PRINT_DESIGN;
    /**
     * 强制分页
     * @type {?}
     */
    Lodop.prototype.NEWPAGE;
    /**
     * 获得打印设备个数
     * @type {?}
     */
    Lodop.prototype.GET_PRINTER_COUNT;
    /**
     * 获得打印设备名称
     * @type {?}
     */
    Lodop.prototype.GET_PRINTER_NAME;
    /**
     * 指定打印设备
     * @type {?}
     */
    Lodop.prototype.SET_PRINTER_INDEX;
    /**
     * 【CLodop】指定打印机
     * @type {?}
     */
    Lodop.prototype.SET_PRINTER_INDEX;
    /**
     * 选择打印设备
     * @type {?}
     */
    Lodop.prototype.SELECT_PRINTER;
    /**
     * 设置显示模式
     * @type {?}
     */
    Lodop.prototype.SET_SHOW_MODE;
    /**
     * 设置打印模式
     * @type {?}
     */
    Lodop.prototype.SET_PRINT_MODE;
    /**
     * 设置打印份数
     * @type {?}
     */
    Lodop.prototype.SET_PRINT_COPIES;
    /**
     * 设置预览窗口
     * @type {?}
     */
    Lodop.prototype.SET_PREVIEW_WINDOW;
    /**
     * 指定背景图
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_SETUP_BKIMG;
    /**
     * 发送原始数据
     * @type {?}
     */
    Lodop.prototype.SEND_PRINT_RAWDATA;
    /**
     * 写端口数据
     * @type {?}
     */
    Lodop.prototype.WRITE_PORT_DATA;
    /**
     * 读端口数据
     * @type {?}
     */
    Lodop.prototype.READ_PORT_DATA;
    /**
     * 获得配置文件名
     * @type {?}
     */
    Lodop.prototype.GET_PRINT_INIFFNAME;
    /**
     * 获得纸张类型名清单
     * @type {?}
     */
    Lodop.prototype.GET_PAGESIZES_LIST;
    /**
     * 写本地文件内容
     * @type {?}
     */
    Lodop.prototype.WRITE_FILE_TEXT;
    /**
     * 读本地文件内容
     * @type {?}
     */
    Lodop.prototype.GET_FILE_TEXT;
    /**
     * 读本地文件时间
     * @type {?}
     */
    Lodop.prototype.GET_FILE_TIME;
    /**
     * 判断本地文件是否存在
     * @type {?}
     */
    Lodop.prototype.IS_FILE_EXIST;
    /**
     * 获得系统信息
     * @type {?}
     */
    Lodop.prototype.GET_SYSTEM_INFO;
    /**
     * 获得数据值
     * @type {?}
     */
    Lodop.prototype.GET_VALUE;
    /**
     * 数据格式转换
     * @type {?}
     */
    Lodop.prototype.FORMAT;
    /**
     * 获得对话框结果值
     * @type {?}
     */
    Lodop.prototype.GET_DIALOG_VALUE;
    /**
     * (增强型)打印初始化
     * @type {?}
     */
    Lodop.prototype.PRINT_INITA;
    /**
     * (增强型)增加超文本打印项(图形模式)
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_HTML;
    /**
     * (增强型)增加表格打印项（URL模式）
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_TBURL;
    /**
     * (增强型)增加纯文本打印项
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_TEXTA;
    /**
     * (增强型)设置打印项风格A, 继承 `SET_PRINT_STYLE` 的所有属性
     * @type {?}
     */
    Lodop.prototype.SET_PRINT_STYLEA;
    /**
     * (增强型)导出数据到文件
     * @type {?}
     */
    Lodop.prototype.SAVE_TO_FILE;
    /**
     * (增强型)设置保存模式
     * @type {?}
     */
    Lodop.prototype.SET_SAVE_MODE;
    /**
     * (增强型)增加图形
     * @type {?}
     */
    Lodop.prototype.ADD_PRINT_SHAPE;
    /**
     * (增强型)指定打印设备
     * @type {?}
     */
    Lodop.prototype.SET_PRINTER_INDEXA;
    /**
     * (增强型)强制分页
     * @type {?}
     */
    Lodop.prototype.NEWPAGEA;
    /**
     * (增强型)打印预览A
     * @type {?}
     */
    Lodop.prototype.PREVIEWA;
    /**
     * (增强型)打印预览B
     * @type {?}
     */
    Lodop.prototype.PREVIEWB;
    /**
     * 直接打印A
     * @type {?}
     */
    Lodop.prototype.PRINTA;
    /**
     * 直接打印B
     * @type {?}
     */
    Lodop.prototype.PRINTB;
    /**
     * 显示图表
     * @type {?}
     */
    Lodop.prototype.SHOW_CHART;
    /**
     * 控制界面动作
     * @type {?}
     */
    Lodop.prototype.DO_ACTION;
    /**
     * 设置软件产品注册信息
     *
     * \@param strCompanyName 注册单位名称，用途与控件参数CompanyName一样。
     * \@param strLicense 主注册号，用途与控件参数License一样。
     * \@param strLicenseA 附加注册号A，用途与控件参数LicenseA一样。
     * \@param strLicenseB 附加注册号B，用途与控件参数LicenseB一样。
     * @type {?}
     */
    Lodop.prototype.SET_LICENSES;
    /** @type {?} */
    Lodop.prototype.webskt;
}
/**
 * @record
 */
export function LodopResult() { }
function LodopResult_tsickle_Closure_declarations() {
    /**
     * 是否成功
     * @type {?}
     */
    LodopResult.prototype.ok;
    /**
     * 错误码
     * @type {?|undefined}
     */
    LodopResult.prototype.status;
    /**
     * 成功时携带 LODOP 对象
     * @type {?|undefined}
     */
    LodopResult.prototype.lodop;
    /**
     * 错误信息
     * @type {?|undefined}
     */
    LodopResult.prototype.error;
}
/**
 * @record
 */
export function LodopPrintResult() { }
function LodopPrintResult_tsickle_Closure_declarations() {
    /**
     * 是否成功
     * @type {?}
     */
    LodopPrintResult.prototype.ok;
    /**
     * 错误信息
     * @type {?|undefined}
     */
    LodopPrintResult.prototype.error;
    /**
     * 代码
     * @type {?}
     */
    LodopPrintResult.prototype.code;
    /**
     * 动态参数上下文对象
     * @type {?}
     */
    LodopPrintResult.prototype.item;
    /**
     * 代码解析表达式
     * @type {?|undefined}
     */
    LodopPrintResult.prototype.parser;
}
