// @delon/abc/src/charts/chart/chart.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Input, EventEmitter, Output, HostBinding, } from '@angular/core';
import { debounceTime } from 'rxjs/operators';
import { FromEventObservable } from 'rxjs/observable/FromEventObservable';
import { coerceNumberProperty } from '@angular/cdk/coercion';
var G2ChartComponent = /** @class */ (function () {
    // endregion
    function G2ChartComponent(el) {
        this.el = el;
        this._resizeTime = 0;
        this.render = new EventEmitter();
        this.resize = new EventEmitter();
        this.destroy = new EventEmitter();
        this.resize$ = null;
    }
    Object.defineProperty(G2ChartComponent.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this._height;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._height = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(G2ChartComponent.prototype, "resizeTime", {
        get: /**
         * @return {?}
         */
        function () {
            return this._resizeTime;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._resizeTime = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    G2ChartComponent.prototype.renderChart = /**
     * @return {?}
     */
    function () {
        this.el.nativeElement.innerHTML = '';
        this.render.emit(this.el);
    };
    /**
     * @return {?}
     */
    G2ChartComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () { return _this.renderChart(); }, 200);
    };
    /**
     * @return {?}
     */
    G2ChartComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy.emit(this.el);
        this.uninstallResizeEvent();
    };
    /**
     * @return {?}
     */
    G2ChartComponent.prototype.installResizeEvent = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.resizeTime <= 0 || !this.resize$)
            return;
        if (this.resizeTime <= 200)
            this.resizeTime = 200;
        this.resize$ = /** @type {?} */ (FromEventObservable.create(window, 'resize')
            .pipe(debounceTime(this.resizeTime))
            .subscribe(function () { return _this.resize.emit(_this.el); }));
    };
    /**
     * @return {?}
     */
    G2ChartComponent.prototype.uninstallResizeEvent = /**
     * @return {?}
     */
    function () {
        if (this.resize$)
            this.resize$.unsubscribe();
    };
    G2ChartComponent.decorators = [
        { type: Component, args: [{
                    selector: 'g2-chart',
                    template: "",
                },] },
    ];
    // endregion
    /** @nocollapse */
    G2ChartComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    G2ChartComponent.propDecorators = {
        "height": [{ type: HostBinding, args: ['style.height.px',] }, { type: Input },],
        "resizeTime": [{ type: Input },],
        "render": [{ type: Output },],
        "resize": [{ type: Output },],
        "destroy": [{ type: Output },],
    };
    return G2ChartComponent;
}());
export { G2ChartComponent };
function G2ChartComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    G2ChartComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    G2ChartComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    G2ChartComponent.propDecorators;
    /** @type {?} */
    G2ChartComponent.prototype._height;
    /** @type {?} */
    G2ChartComponent.prototype._resizeTime;
    /** @type {?} */
    G2ChartComponent.prototype.render;
    /** @type {?} */
    G2ChartComponent.prototype.resize;
    /** @type {?} */
    G2ChartComponent.prototype.destroy;
    /** @type {?} */
    G2ChartComponent.prototype.resize$;
    /** @type {?} */
    G2ChartComponent.prototype.el;
}
