// @delon/abc/src/global-footer/global-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
var GlobalFooterComponent = /** @class */ (function () {
    function GlobalFooterComponent() {
    }
    GlobalFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'global-footer',
                    template: "\n  <div *ngIf=\"links && links.length > 0\" class=\"links\">\n    <a *ngFor=\"let i of links\" routerLink=\"{{i.href}}\" [attr.target]=\"i.blankTarget\">{{i.title}}</a>\n  </div>\n  <div class=\"copyright\"><ng-content></ng-content></div>\n  ",
                    host: { '[class.ad-global-footer]': 'true' },
                    preserveWhitespaces: false,
                },] },
    ];
    /** @nocollapse */
    GlobalFooterComponent.propDecorators = {
        "links": [{ type: Input },],
    };
    return GlobalFooterComponent;
}());
export { GlobalFooterComponent };
function GlobalFooterComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GlobalFooterComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GlobalFooterComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GlobalFooterComponent.propDecorators;
    /** @type {?} */
    GlobalFooterComponent.prototype.links;
}
