// @delon/abc/src/trend/trend.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
var TrendComponent = /** @class */ (function () {
    function TrendComponent() {
        this._colorful = true;
    }
    Object.defineProperty(TrendComponent.prototype, "colorful", {
        get: /**
         * 是否彩色标记
         * @return {?}
         */
        function () {
            return this._colorful;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._colorful = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    TrendComponent.decorators = [
        { type: Component, args: [{
                    selector: 'trend',
                    template: "\n  <ng-content></ng-content>\n  <span *ngIf=\"flag\" class=\"{{flag}}\"><i class=\"anticon anticon-caret-{{flag}}\"></i></span>\n  ",
                    host: {
                        '[class.ad-trend]': 'true',
                        '[class.grey]': '!colorful',
                    },
                    preserveWhitespaces: false,
                },] },
    ];
    /** @nocollapse */
    TrendComponent.propDecorators = {
        "flag": [{ type: Input },],
        "colorful": [{ type: Input },],
    };
    return TrendComponent;
}());
export { TrendComponent };
function TrendComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TrendComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TrendComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    TrendComponent.propDecorators;
    /**
     * 上升下降标识
     * @type {?}
     */
    TrendComponent.prototype.flag;
    /** @type {?} */
    TrendComponent.prototype._colorful;
}
