// @delon/abc/src/charts/bar/bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, HostBinding, ViewChild, ElementRef, TemplateRef, ChangeDetectionStrategy, ChangeDetectorRef, } from '@angular/core';
import { debounceTime } from 'rxjs/operators';
import { FromEventObservable } from 'rxjs/observable/FromEventObservable';
import { coerceBooleanProperty, coerceNumberProperty, } from '@angular/cdk/coercion';
var G2BarComponent = /** @class */ (function () {
    function G2BarComponent(el, cd) {
        this.el = el;
        this.cd = cd;
        // region: fields
        this._title = '';
        this.color = 'rgba(24, 144, 255, 0.85)';
        this._height = 0;
        this._autoLabel = true;
        // endregion
        this._cls = true;
        this.autoHideXLabels = false;
        this.scroll$ = null;
        this.cd.detach();
    }
    Object.defineProperty(G2BarComponent.prototype, "title", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this._title = null;
                this._titleTpl = value;
            }
            else
                this._title = value;
            this.cd.detectChanges();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(G2BarComponent.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this._height;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._height = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(G2BarComponent.prototype, "autoLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return this._autoLabel;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._autoLabel = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    G2BarComponent.prototype.install = /**
     * @return {?}
     */
    function () {
        if (!this.data || (this.data && this.data.length < 1))
            return;
        // this.uninstall();
        this.node.nativeElement.innerHTML = '';
        var /** @type {?} */ padding = Object.assign([], this.padding);
        if (padding.length <= 0)
            padding = [32, 0, this.autoHideXLabels ? 8 : 32, 40];
        var /** @type {?} */ chart = new G2.Chart({
            container: this.node.nativeElement,
            forceFit: true,
            height: +this.height - 22,
            legend: null,
            padding: padding,
        });
        chart.axis('x', !this.autoHideXLabels);
        chart.axis('y', {
            title: false,
            line: false,
            tickLine: false,
        });
        chart.source(this.data, {
            x: {
                type: 'cat',
            },
            y: {
                min: 0,
            },
        });
        chart.tooltip({
            showTitle: false,
        });
        chart
            .interval()
            .position('x*y')
            .color(this.color)
            .tooltip('x*y', function (x, y) {
            return {
                name: x,
                value: y,
            };
        });
        chart.render();
        setTimeout(function () {
            chart.forceFit();
            chart.repaint();
        }, 30);
        this.chart = chart;
    };
    /**
     * @return {?}
     */
    G2BarComponent.prototype.uninstall = /**
     * @return {?}
     */
    function () {
        if (this.chart) {
            this.chart.destroy();
            this.chart = null;
        }
    };
    /**
     * @return {?}
     */
    G2BarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.installResizeEvent();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    G2BarComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.resize();
    };
    /**
     * @return {?}
     */
    G2BarComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.uninstallResizeEvent();
        this.uninstall();
    };
    /**
     * @return {?}
     */
    G2BarComponent.prototype.installResizeEvent = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.autoLabel || this.scroll$)
            return;
        this.scroll$ = /** @type {?} */ (FromEventObservable.create(window, 'resize')
            .pipe(debounceTime(200))
            .subscribe(function () { return _this.resize(); }));
    };
    /**
     * @return {?}
     */
    G2BarComponent.prototype.uninstallResizeEvent = /**
     * @return {?}
     */
    function () {
        if (this.scroll$)
            this.scroll$.unsubscribe();
    };
    /**
     * @return {?}
     */
    G2BarComponent.prototype.resize = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ canvasWidth = this.el.nativeElement.clientWidth;
        var /** @type {?} */ minWidth = this.data.length * 30;
        if (canvasWidth <= minWidth) {
            if (!this.autoHideXLabels) {
                this.autoHideXLabels = true;
                this.install();
                return;
            }
        }
        else if (this.autoHideXLabels) {
            this.autoHideXLabels = false;
            this.install();
            return;
        }
        if (!this.chart)
            this.install();
    };
    G2BarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'g2-bar',
                    template: "\n  <ng-container *ngIf=\"_title; else _titleTpl\"><h4>{{_title}}</h4></ng-container>\n  <div #container></div>",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                },] },
    ];
    // endregion
    /** @nocollapse */
    G2BarComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ChangeDetectorRef, },
    ]; };
    G2BarComponent.propDecorators = {
        "title": [{ type: Input },],
        "color": [{ type: Input },],
        "height": [{ type: HostBinding, args: ['style.height.px',] }, { type: Input },],
        "padding": [{ type: Input },],
        "data": [{ type: Input },],
        "autoLabel": [{ type: Input },],
        "_cls": [{ type: HostBinding, args: ['class.g2-chart',] },],
        "node": [{ type: ViewChild, args: ['container',] },],
    };
    return G2BarComponent;
}());
export { G2BarComponent };
function G2BarComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    G2BarComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    G2BarComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    G2BarComponent.propDecorators;
    /** @type {?} */
    G2BarComponent.prototype._title;
    /** @type {?} */
    G2BarComponent.prototype._titleTpl;
    /** @type {?} */
    G2BarComponent.prototype.color;
    /** @type {?} */
    G2BarComponent.prototype._height;
    /** @type {?} */
    G2BarComponent.prototype.padding;
    /** @type {?} */
    G2BarComponent.prototype.data;
    /** @type {?} */
    G2BarComponent.prototype._autoLabel;
    /** @type {?} */
    G2BarComponent.prototype._cls;
    /** @type {?} */
    G2BarComponent.prototype.node;
    /** @type {?} */
    G2BarComponent.prototype.chart;
    /** @type {?} */
    G2BarComponent.prototype.autoHideXLabels;
    /** @type {?} */
    G2BarComponent.prototype.scroll$;
    /** @type {?} */
    G2BarComponent.prototype.el;
    /** @type {?} */
    G2BarComponent.prototype.cd;
}
