// @delon/abc/src/exception/exception.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ContentChild, TemplateRef } from '@angular/core';
var ExceptionComponent = /** @class */ (function () {
    function ExceptionComponent() {
        this._img = '';
        this._title = '';
        this._desc = '';
    }
    Object.defineProperty(ExceptionComponent.prototype, "type", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var /** @type {?} */ item = {
                403: {
                    img: 'https://gw.alipayobjects.com/zos/rmsportal/wZcnGqRDyhPOEYFcZDnb.svg',
                    title: '403',
                    desc: '抱歉，你无权访问该页面',
                },
                404: {
                    img: 'https://gw.alipayobjects.com/zos/rmsportal/KpnpchXsobRgLElEozzI.svg',
                    title: '404',
                    desc: '抱歉，你访问的页面不存在',
                },
                500: {
                    img: 'https://gw.alipayobjects.com/zos/rmsportal/RVRUAYdCGeYNBWoKiIwB.svg',
                    title: '500',
                    desc: '抱歉，服务器出错了',
                },
            }[value];
            if (!item)
                return;
            this._img = item.img;
            this._title = item.title;
            this._desc = item.desc;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExceptionComponent.prototype, "img", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._img = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExceptionComponent.prototype, "title", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExceptionComponent.prototype, "desc", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._desc = value;
        },
        enumerable: true,
        configurable: true
    });
    ExceptionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'exception',
                    template: "\n  <div class=\"img-block\">\n    <div class=\"img\" [ngStyle]=\"{'background-image': 'url(' + _img + ')'}\"></div>\n  </div>\n  <div class=\"cont\">\n    <h1 [innerHTML]=\"_title\"></h1>\n    <div class=\"desc\" [innerHTML]=\"_desc\"></div>\n    <ng-template #defaultActions>\n      <button nz-button [routerLink]=\"['/']\" [nzType]=\"'primary'\">\u8FD4\u56DE\u9996\u9875</button>\n      <ng-content></ng-content>\n    </ng-template>\n    <div class=\"actions\" *ngIf=\"actions; else defaultActions\">\n      <ng-template [ngTemplateOutlet]=\"actions\"></ng-template>\n      <ng-content></ng-content>\n    </div>\n  </div>\n  ",
                    host: { '[class.ad-exception]': 'true' },
                    preserveWhitespaces: false,
                },] },
    ];
    /** @nocollapse */
    ExceptionComponent.propDecorators = {
        "type": [{ type: Input },],
        "img": [{ type: Input },],
        "title": [{ type: Input },],
        "desc": [{ type: Input },],
        "actions": [{ type: ContentChild, args: ['actions',] },],
    };
    return ExceptionComponent;
}());
export { ExceptionComponent };
function ExceptionComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ExceptionComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ExceptionComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    ExceptionComponent.propDecorators;
    /** @type {?} */
    ExceptionComponent.prototype._img;
    /** @type {?} */
    ExceptionComponent.prototype._title;
    /** @type {?} */
    ExceptionComponent.prototype._desc;
    /** @type {?} */
    ExceptionComponent.prototype.actions;
}
