// @delon/mock/src/src/interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function MockCachedRule() { }
function MockCachedRule_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [key: string]: any;
    */
    /** @type {?} */
    MockCachedRule.prototype.method;
    /** @type {?} */
    MockCachedRule.prototype.url;
    /** @type {?} */
    MockCachedRule.prototype.martcher;
    /** @type {?} */
    MockCachedRule.prototype.segments;
    /** @type {?} */
    MockCachedRule.prototype.callback;
}
/**
 * @record
 */
export function MockRule() { }
function MockRule_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [key: string]: any;
    */
    /** @type {?} */
    MockRule.prototype.method;
    /** @type {?} */
    MockRule.prototype.url;
    /**
     * 路由参数
     * @type {?|undefined}
     */
    MockRule.prototype.params;
    /** @type {?} */
    MockRule.prototype.callback;
}
/**
 * @record
 */
export function MockRequest() { }
function MockRequest_tsickle_Closure_declarations() {
    /**
     * 路由参数
     * @type {?|undefined}
     */
    MockRequest.prototype.params;
    /**
     * URL参数
     * @type {?|undefined}
     */
    MockRequest.prototype.queryString;
    /** @type {?|undefined} */
    MockRequest.prototype.headers;
    /** @type {?|undefined} */
    MockRequest.prototype.body;
    /**
     * 原始 `HttpRequest`
     * @type {?}
     */
    MockRequest.prototype.original;
}
