// @delon/abc/src/full-content/full-content.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FullContentComponent } from './full-content.component';
import { FullContentToggleDirective } from './full-content-toggle.directive';
import { FullContentService } from './full-content.service';
var /** @type {?} */ COMPONENTS = [FullContentComponent, FullContentToggleDirective];
var AdFullContentModule = /** @class */ (function () {
    function AdFullContentModule() {
    }
    /**
     * @return {?}
     */
    AdFullContentModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AdFullContentModule,
            providers: [FullContentService],
        };
    };
    AdFullContentModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice(),
                },] },
    ];
    return AdFullContentModule;
}());
export { AdFullContentModule };
function AdFullContentModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AdFullContentModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AdFullContentModule.ctorParameters;
}
