// @delon/abc/src/full-content/full-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, ChangeDetectionStrategy, ChangeDetectorRef, Input, Output, EventEmitter, Inject, HostBinding, } from '@angular/core';
import { coerceBooleanProperty, coerceNumberProperty, } from '@angular/cdk/coercion';
import { DOCUMENT } from '@angular/platform-browser';
import { FromEventObservable } from 'rxjs/observable/FromEventObservable';
import { debounceTime, filter } from 'rxjs/operators';
import { FullContentService } from './full-content.service';
import { Router, ActivationStart, ActivationEnd } from '@angular/router';
var /** @type {?} */ cls = "ad-full-content-wrap";
var /** @type {?} */ fsCls = "ad-full-content-fs";
var /** @type {?} */ hideTitleCls = "ad-full-content-ht";
var FullContentComponent = /** @class */ (function () {
    // endregion
    function FullContentComponent(el, router, cd, srv, doc) {
        this.el = el;
        this.router = router;
        this.cd = cd;
        this.srv = srv;
        this.doc = doc;
        this.inited = false;
        this.id = "_full-content-" + Math.random()
            .toString(36)
            .substring(2);
        this._height = 0;
        this._hideTitle = true;
        this._padding = 24;
        this.fullscreenChange = new EventEmitter();
        this.scroll$ = null;
    }
    Object.defineProperty(FullContentComponent.prototype, "fullscreen", {
        get: /**
         * @return {?}
         */
        function () {
            return this._fullscreen;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._fullscreen = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FullContentComponent.prototype, "hideTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._hideTitle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._hideTitle = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FullContentComponent.prototype, "padding", {
        get: /**
         * @return {?}
         */
        function () {
            return this._padding;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._padding = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FullContentComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.inited = true;
        this.bodyEl = this.doc.querySelector('body');
        this.bodyEl.classList.add(cls);
        (/** @type {?} */ (this.el.nativeElement)).id = this.id;
        this.update();
        this.installResizeEvent();
        this.srv$ = /** @type {?} */ (this.srv.change.subscribe(function (res) {
            if (res)
                _this.toggle();
        }));
        this.route$ = /** @type {?} */ (this.router.events
            .pipe(filter(function (e) {
            return e instanceof ActivationStart || e instanceof ActivationEnd;
        }), debounceTime(200))
            .subscribe(function (e) {
            if (!!document.querySelector('#' + _this.id)) {
                _this.bodyEl.classList.add(cls);
                _this.updateFsCls();
            }
            else {
                _this.bodyEl.classList.remove(cls, fsCls, hideTitleCls);
            }
        }));
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () { return _this.updateHeight(); });
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.updateFsCls = /**
     * @return {?}
     */
    function () {
        if (this.fullscreen) {
            this.bodyEl.classList.add(fsCls);
            if (this.hideTitle)
                this.bodyEl.classList.add(hideTitleCls);
        }
        else {
            this.bodyEl.classList.remove(fsCls);
            if (this.hideTitle)
                this.bodyEl.classList.remove(hideTitleCls);
        }
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.update = /**
     * @return {?}
     */
    function () {
        this.updateFsCls();
        this.fullscreenChange.emit(this.fullscreen);
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.updateHeight = /**
     * @return {?}
     */
    function () {
        this._height =
            this.bodyEl.getBoundingClientRect().height -
                (/** @type {?} */ (this.el.nativeElement)).getBoundingClientRect().top -
                this.padding;
        this.cd.detectChanges();
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.toggle = /**
     * @return {?}
     */
    function () {
        this.fullscreen = !this.fullscreen;
        this.update();
        this.updateHeight();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    FullContentComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.inited)
            this.update();
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.bodyEl.classList.remove(cls, fsCls, hideTitleCls);
        this.uninstallResizeEvent();
        this.srv$.unsubscribe();
        this.route$.unsubscribe();
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.installResizeEvent = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.scroll$ = /** @type {?} */ (FromEventObservable.create(window, 'resize')
            .pipe(debounceTime(200))
            .subscribe(function () { return _this.updateHeight(); }));
    };
    /**
     * @return {?}
     */
    FullContentComponent.prototype.uninstallResizeEvent = /**
     * @return {?}
     */
    function () {
        this.scroll$.unsubscribe();
    };
    FullContentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'full-content',
                    template: "<ng-content></ng-content>",
                    host: { '[class.ad-full-content]': 'true' },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                },] },
    ];
    // endregion
    /** @nocollapse */
    FullContentComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Router, },
        { type: ChangeDetectorRef, },
        { type: FullContentService, },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
    ]; };
    FullContentComponent.propDecorators = {
        "_height": [{ type: HostBinding, args: ['style.height.px',] },],
        "fullscreen": [{ type: Input },],
        "hideTitle": [{ type: Input },],
        "padding": [{ type: Input },],
        "fullscreenChange": [{ type: Output },],
    };
    return FullContentComponent;
}());
export { FullContentComponent };
function FullContentComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    FullContentComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    FullContentComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    FullContentComponent.propDecorators;
    /** @type {?} */
    FullContentComponent.prototype.bodyEl;
    /** @type {?} */
    FullContentComponent.prototype.inited;
    /** @type {?} */
    FullContentComponent.prototype.srv$;
    /** @type {?} */
    FullContentComponent.prototype.route$;
    /** @type {?} */
    FullContentComponent.prototype.id;
    /** @type {?} */
    FullContentComponent.prototype._height;
    /** @type {?} */
    FullContentComponent.prototype._fullscreen;
    /** @type {?} */
    FullContentComponent.prototype._hideTitle;
    /** @type {?} */
    FullContentComponent.prototype._padding;
    /** @type {?} */
    FullContentComponent.prototype.fullscreenChange;
    /** @type {?} */
    FullContentComponent.prototype.scroll$;
    /** @type {?} */
    FullContentComponent.prototype.el;
    /** @type {?} */
    FullContentComponent.prototype.router;
    /** @type {?} */
    FullContentComponent.prototype.cd;
    /** @type {?} */
    FullContentComponent.prototype.srv;
    /** @type {?} */
    FullContentComponent.prototype.doc;
}
