// ngx-notify/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NotifyPushService } from './components/push.service';
import { NotifyComponent } from './components/notify.component';
import { NotifyHolderComponent } from './components/notify-holder.component';
import { NotifyService } from './components/notify.service';
import { NotifyServiceConfig } from './components/notify.service.config';
export * from './components/interfaces/icons';
export * from './components/notify-holder.component';
export * from './components/notify.component';
export * from './components/notify.service.config';
export * from './components/notify.service';
export * from './components/push.service';
var NotifyModule = (function () {
    function NotifyModule() {
    }
    NotifyModule.forRoot = function (config) {
        return {
            ngModule: NotifyModule,
            providers: [
                { provide: NotifyServiceConfig, useValue: config }
            ]
        };
    };
    NotifyModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [NotifyHolderComponent, NotifyComponent],
                    providers: [NotifyService, NotifyPushService],
                    entryComponents: [NotifyHolderComponent, NotifyComponent]
                },] },
    ];
    /** @nocollapse */
    NotifyModule.ctorParameters = function () { return []; };
    return NotifyModule;
}());
export { NotifyModule };
