// @delon/abc/src/standard-form-row/standard-form-row.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ElementRef, Renderer2, } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
var StandardFormRowComponent = /** @class */ (function () {
    function StandardFormRowComponent(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this._classMap = [];
    }
    Object.defineProperty(StandardFormRowComponent.prototype, "block", {
        get: /**
         * 是否整行
         * @return {?}
         */
        function () {
            return this._block;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._block = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StandardFormRowComponent.prototype, "last", {
        get: /**
         * 是否最后一行
         * @return {?}
         */
        function () {
            return this._last;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._last = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StandardFormRowComponent.prototype, "grid", {
        get: /**
         * 是否网格布局
         * @return {?}
         */
        function () {
            return this._grid;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._grid = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StandardFormRowComponent.prototype.setClass = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._classMap.forEach(function (cls) {
            return _this.renderer.removeClass(_this.el.nativeElement, cls);
        });
        this._classMap = ["standard-form-row"];
        if (this.block)
            this._classMap.push('block');
        if (this.grid)
            this._classMap.push('grid');
        if (this.last)
            this._classMap.push('last');
        this._classMap.forEach(function (v) {
            return _this.renderer.addClass(_this.el.nativeElement, v);
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    StandardFormRowComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.setClass();
    };
    StandardFormRowComponent.decorators = [
        { type: Component, args: [{
                    selector: 'standard-form-row',
                    template: "\n  <div *ngIf=\"title\" class=\"label\"><span>{{title}}</span></div>\n  <div class=\"control\"><ng-content></ng-content></div>\n  ",
                    host: { '[class.ad-standard-form-row]': 'true' },
                    preserveWhitespaces: false,
                },] },
    ];
    /** @nocollapse */
    StandardFormRowComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    StandardFormRowComponent.propDecorators = {
        "title": [{ type: Input },],
        "block": [{ type: Input },],
        "last": [{ type: Input },],
        "grid": [{ type: Input },],
    };
    return StandardFormRowComponent;
}());
export { StandardFormRowComponent };
function StandardFormRowComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    StandardFormRowComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    StandardFormRowComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    StandardFormRowComponent.propDecorators;
    /** @type {?} */
    StandardFormRowComponent.prototype.title;
    /** @type {?} */
    StandardFormRowComponent.prototype._block;
    /** @type {?} */
    StandardFormRowComponent.prototype._last;
    /** @type {?} */
    StandardFormRowComponent.prototype._grid;
    /** @type {?} */
    StandardFormRowComponent.prototype._classMap;
    /** @type {?} */
    StandardFormRowComponent.prototype.el;
    /** @type {?} */
    StandardFormRowComponent.prototype.renderer;
}
