// @delon/abc/src/error-collect/error-collect.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ErrorCollectComponent } from './error-collect.component';
import { AdErrorCollectConfig } from './error-collect.config';
var /** @type {?} */ COMPONENTS = [ErrorCollectComponent];
var AdErrorCollectModule = /** @class */ (function () {
    function AdErrorCollectModule() {
    }
    /**
     * @return {?}
     */
    AdErrorCollectModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AdErrorCollectModule,
            providers: [AdErrorCollectConfig],
        };
    };
    AdErrorCollectModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice(),
                },] },
    ];
    return AdErrorCollectModule;
}());
export { AdErrorCollectModule };
function AdErrorCollectModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AdErrorCollectModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AdErrorCollectModule.ctorParameters;
}
