// @delon/abc/src/reuse-tab/reuse-tab.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { ReuseTabComponent } from './reuse-tab.component';
import { ReuseTabService } from './reuse-tab.service';
import { ReuseTabContextComponent } from './reuse-tab-context.component';
import { ReuseTabContextDirective } from './reuse-tab-context.directive';
import { ReuseTabContextMenuComponent } from './reuse-tab-context-menu.component';
var /** @type {?} */ COMPONENTS = [ReuseTabComponent];
var /** @type {?} */ NOEXPORTS = [
    ReuseTabContextMenuComponent,
    ReuseTabContextComponent,
    ReuseTabContextDirective,
];
import { OverlayModule } from '@angular/cdk/overlay';
import { NgZorroAntdModule } from 'ng-zorro-antd';
var AdReuseTabModule = /** @class */ (function () {
    function AdReuseTabModule() {
    }
    /**
     * @return {?}
     */
    AdReuseTabModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AdReuseTabModule,
            providers: [
                ReuseTabService
            ],
        };
    };
    AdReuseTabModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, RouterModule, NgZorroAntdModule, OverlayModule],
                    declarations: COMPONENTS.concat(NOEXPORTS),
                    entryComponents: [ReuseTabContextMenuComponent],
                    exports: COMPONENTS.slice(),
                },] },
    ];
    return AdReuseTabModule;
}());
export { AdReuseTabModule };
function AdReuseTabModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AdReuseTabModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AdReuseTabModule.ctorParameters;
}
