// @delon/abc/src/reuse-tab/reuse-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, ChangeDetectionStrategy, ChangeDetectorRef, EventEmitter, ElementRef, Renderer2, Inject, Optional, } from '@angular/core';
import { Router, NavigationEnd, ActivatedRoute } from '@angular/router';
import { DOCUMENT } from '@angular/platform-browser';
import { combineLatest } from 'rxjs/observable/combineLatest';
import { filter } from 'rxjs/operators';
import { coerceNumberProperty, coerceBooleanProperty, } from '@angular/cdk/coercion';
import { ALAIN_I18N_TOKEN } from '@delon/theme';
import { ReuseTabService } from './reuse-tab.service';
import { ReuseTabMatchMode, } from './interface';
import { ReuseTabContextService } from './reuse-tab-context.service';
var ReuseTabComponent = /** @class */ (function () {
    // endregion
    function ReuseTabComponent(srv, cd, router, route, el, render, doc, i18nSrv) {
        var _this = this;
        this.srv = srv;
        this.cd = cd;
        this.router = router;
        this.route = route;
        this.el = el;
        this.render = render;
        this.doc = doc;
        this.i18nSrv = i18nSrv;
        this.list = [];
        this.pos = 0;
        /**
         * 设置匹配模式
         */
        this.mode = ReuseTabMatchMode.Menu;
        this._debug = false;
        this._allowClose = true;
        this._fixed = true;
        this._showCurrent = true;
        /**
         * 切换时回调
         */
        this.change = new EventEmitter();
        /**
         * 关闭回调
         */
        this.close = new EventEmitter();
        var /** @type {?} */ route$ = this.router.events.pipe(filter(function (evt) { return evt instanceof NavigationEnd; }));
        this.sub$ = /** @type {?} */ (combineLatest(this.srv.change, route$).subscribe(function (_a) {
            var res = _a[0], e = _a[1];
            return _this.genList(res);
        }));
        if (this.i18nSrv)
            this.i18n$ = this.i18nSrv.change.subscribe(function () { return _this.genList(); });
    }
    Object.defineProperty(ReuseTabComponent.prototype, "debug", {
        get: /**
         * 是否Debug模式
         * @return {?}
         */
        function () {
            return this._debug;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._debug = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ReuseTabComponent.prototype, "max", {
        get: /**
         * 允许最多复用多少个页面
         * @return {?}
         */
        function () {
            return this._max;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._max = coerceNumberProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ReuseTabComponent.prototype, "allowClose", {
        get: /**
         * 允许关闭
         * @return {?}
         */
        function () {
            return this._allowClose;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._allowClose = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ReuseTabComponent.prototype, "fixed", {
        get: /**
         * 是否固定
         * @return {?}
         */
        function () {
            return this._fixed;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._fixed = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ReuseTabComponent.prototype, "showCurrent", {
        get: /**
         * 总是显示当前页
         * @return {?}
         */
        function () {
            return this._showCurrent;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._showCurrent = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} title
     * @return {?}
     */
    ReuseTabComponent.prototype.genTit = /**
     * @param {?} title
     * @return {?}
     */
    function (title) {
        return title.i18n && this.i18nSrv
            ? this.i18nSrv.fanyi(title.i18n)
            : title.text;
    };
    /**
     * @param {?=} notify
     * @return {?}
     */
    ReuseTabComponent.prototype.genList = /**
     * @param {?=} notify
     * @return {?}
     */
    function (notify) {
        var _this = this;
        var /** @type {?} */ isClosed = notify && notify.active === 'close';
        var /** @type {?} */ beforeClosePos = isClosed
            ? this.list.findIndex(function (w) { return w.url === notify["url"]; })
            : -1;
        var /** @type {?} */ ls = this.srv.items.map(function (item, index) {
            return /** @type {?} */ ({
                url: item.url,
                title: _this.genTit(item.title),
                closable: _this.allowClose && item.closable && _this.srv.count > 0,
                index: index,
                active: false,
                last: false,
            });
        });
        if (this.showCurrent) {
            var /** @type {?} */ snapshot = this.route.snapshot;
            var /** @type {?} */ url_1 = this.srv.getUrl(snapshot);
            var /** @type {?} */ idx = ls.findIndex(function (w) { return w.url === url_1; });
            // jump directly when the current exists in the list
            // or create a new current item and jump
            if (idx !== -1 || (isClosed && notify["url"] === url_1)) {
                this.pos = isClosed
                    ? idx >= beforeClosePos
                        ? this.pos - 1
                        : this.pos
                    : idx;
            }
            else {
                var /** @type {?} */ snapshotTrue = this.srv.getTruthRoute(snapshot);
                ls.push(/** @type {?} */ ({
                    url: url_1,
                    title: this.genTit(this.srv.getTitle(url_1, snapshotTrue)),
                    closable: this.allowClose &&
                        this.srv.count > 0 &&
                        this.srv.getClosable(url_1, snapshotTrue),
                    index: ls.length,
                    active: false,
                    last: false,
                }));
                this.pos = ls.length - 1;
            }
            // fix unabled close last item
            if (ls.length <= 1)
                ls[0].closable = false;
        }
        this.list = ls;
        if (ls.length && isClosed) {
            this.to(null, this.pos);
        }
        this.refStatus(false);
        this.visibility();
        this.cd.detectChanges();
    };
    /**
     * @return {?}
     */
    ReuseTabComponent.prototype.visibility = /**
     * @return {?}
     */
    function () {
        if (this.showCurrent)
            return;
        this.render.setStyle(this.el.nativeElement, 'display', this.list.length === 0 ? 'none' : 'block');
    };
    // region: UI
    /**
     * @param {?} res
     * @return {?}
     */
    ReuseTabComponent.prototype.cmChange = /**
     * @param {?} res
     * @return {?}
     */
    function (res) {
        switch (res.type) {
            case 'close':
                this._close(null, res.item.index, res.includeNonCloseable);
                break;
            case 'closeRight':
                this.srv.closeRight(res.item.url, res.includeNonCloseable);
                this.close.emit(null);
                break;
            case 'clear':
            case 'closeOther':
                this.srv.clear(res.includeNonCloseable);
                this.close.emit(null);
                break;
        }
    };
    /**
     * @param {?=} dc
     * @return {?}
     */
    ReuseTabComponent.prototype.refStatus = /**
     * @param {?=} dc
     * @return {?}
     */
    function (dc) {
        var _this = this;
        if (dc === void 0) { dc = true; }
        if (this.list.length) {
            this.list[this.list.length - 1].last = true;
            this.list.forEach(function (i, idx) { return (i.active = _this.pos === idx); });
        }
        if (dc)
            this.cd.detectChanges();
    };
    /**
     * @param {?} e
     * @param {?} index
     * @return {?}
     */
    ReuseTabComponent.prototype.to = /**
     * @param {?} e
     * @param {?} index
     * @return {?}
     */
    function (e, index) {
        var _this = this;
        if (e) {
            e.preventDefault();
            e.stopPropagation();
        }
        index = Math.max(0, Math.min(index, this.list.length - 1));
        var /** @type {?} */ item = this.list[index];
        this.router.navigateByUrl(item.url).then(function (res) {
            if (!res)
                return;
            _this.pos = index;
            _this.item = item;
            _this.refStatus();
            _this.change.emit(item);
        });
    };
    /**
     * @param {?} e
     * @param {?} idx
     * @param {?} includeNonCloseable
     * @return {?}
     */
    ReuseTabComponent.prototype._close = /**
     * @param {?} e
     * @param {?} idx
     * @param {?} includeNonCloseable
     * @return {?}
     */
    function (e, idx, includeNonCloseable) {
        if (e) {
            e.preventDefault();
            e.stopPropagation();
        }
        var /** @type {?} */ item = this.list[idx];
        this.srv.close(item.url, includeNonCloseable);
        this.close.emit(item);
        this.cd.detectChanges();
        return false;
    };
    // endregion
    /**
     * @return {?}
     */
    ReuseTabComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setClass();
        this.genList();
    };
    /**
     * @return {?}
     */
    ReuseTabComponent.prototype.setClass = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ el = this.el.nativeElement;
        var /** @type {?} */ body = this.doc.querySelector('body');
        var /** @type {?} */ fixedCls = "fixed";
        var /** @type {?} */ bodyCls = "has-reuse-tab";
        if (this.fixed) {
            this.render.addClass(el, fixedCls);
            this.render.addClass(body, bodyCls);
        }
        else {
            this.render.removeClass(el, fixedCls);
            this.render.removeClass(body, bodyCls);
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ReuseTabComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.max)
            this.srv.max = this.max;
        if (changes.excludes)
            this.srv.excludes = this.excludes;
        if (changes.mode)
            this.srv.mode = this.mode;
        this.srv.debug = this.debug;
        this.setClass();
        this.cd.detectChanges();
    };
    /**
     * @return {?}
     */
    ReuseTabComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.sub$.unsubscribe();
        if (this.i18n$)
            this.i18n$.unsubscribe();
    };
    ReuseTabComponent.decorators = [
        { type: Component, args: [{
                    selector: 'reuse-tab',
                    template: "<nz-tabset [nzSelectedIndex]=\"pos\" [nzAnimated]=\"false\" nzType=\"line\"> <nz-tab *ngFor=\"let i of list; let index = index\" [nzTitle]=\"titleTemplate\"> <ng-template #titleTemplate> <span [context-menu]=\"i\" (click)=\"to($event, index)\" class=\"name\">{{i.title}}</span> <i *ngIf=\"i.closable\" class=\"anticon anticon-close op\" (click)=\"_close($event, index, false)\"></i> </ng-template> </nz-tab> </nz-tabset> <reuse-tab-context [i18n]=\"i18n\" (change)=\"cmChange($event)\"></reuse-tab-context> ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    providers: [ReuseTabContextService],
                },] },
    ];
    /** @nocollapse */
    ReuseTabComponent.ctorParameters = function () { return [
        { type: ReuseTabService, },
        { type: ChangeDetectorRef, },
        { type: Router, },
        { type: ActivatedRoute, },
        { type: ElementRef, },
        { type: Renderer2, },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [ALAIN_I18N_TOKEN,] },] },
    ]; };
    ReuseTabComponent.propDecorators = {
        "mode": [{ type: Input },],
        "i18n": [{ type: Input },],
        "debug": [{ type: Input },],
        "max": [{ type: Input },],
        "excludes": [{ type: Input },],
        "allowClose": [{ type: Input },],
        "fixed": [{ type: Input },],
        "showCurrent": [{ type: Input },],
        "change": [{ type: Output },],
        "close": [{ type: Output },],
    };
    return ReuseTabComponent;
}());
export { ReuseTabComponent };
function ReuseTabComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ReuseTabComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ReuseTabComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    ReuseTabComponent.propDecorators;
    /** @type {?} */
    ReuseTabComponent.prototype.sub$;
    /** @type {?} */
    ReuseTabComponent.prototype.i18n$;
    /** @type {?} */
    ReuseTabComponent.prototype.list;
    /** @type {?} */
    ReuseTabComponent.prototype.item;
    /** @type {?} */
    ReuseTabComponent.prototype.pos;
    /**
     * 设置匹配模式
     * @type {?}
     */
    ReuseTabComponent.prototype.mode;
    /**
     * 选项文本国际化
     * @type {?}
     */
    ReuseTabComponent.prototype.i18n;
    /** @type {?} */
    ReuseTabComponent.prototype._debug;
    /** @type {?} */
    ReuseTabComponent.prototype._max;
    /**
     * 排除规则，限 `mode=URL`
     * @type {?}
     */
    ReuseTabComponent.prototype.excludes;
    /** @type {?} */
    ReuseTabComponent.prototype._allowClose;
    /** @type {?} */
    ReuseTabComponent.prototype._fixed;
    /** @type {?} */
    ReuseTabComponent.prototype._showCurrent;
    /**
     * 切换时回调
     * @type {?}
     */
    ReuseTabComponent.prototype.change;
    /**
     * 关闭回调
     * @type {?}
     */
    ReuseTabComponent.prototype.close;
    /** @type {?} */
    ReuseTabComponent.prototype.srv;
    /** @type {?} */
    ReuseTabComponent.prototype.cd;
    /** @type {?} */
    ReuseTabComponent.prototype.router;
    /** @type {?} */
    ReuseTabComponent.prototype.route;
    /** @type {?} */
    ReuseTabComponent.prototype.el;
    /** @type {?} */
    ReuseTabComponent.prototype.render;
    /** @type {?} */
    ReuseTabComponent.prototype.doc;
    /** @type {?} */
    ReuseTabComponent.prototype.i18nSrv;
}
