// @delon/form/src/src/widgets/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export { ObjectWidget } from './object/object.widget';
export { ArrayWidget } from './array/array.widget';
export { StringWidget } from './string/string.widget';
export { NumberWidget } from './number/number.widget';
export { DateWidget } from './date/date.widget';
export { TimeWidget } from './time/time.widget';
export { RadioWidget } from './radio/radio.widget';
export { CheckboxWidget } from './checkbox/checkbox.widget';
export { BooleanWidget } from './boolean/boolean.widget';
export { TextareaWidget } from './textarea/textarea.widget';
export { SelectWidget } from './select/select.widget';
export { TagWidget } from './tag/tag.widget';
export { UploadWidget } from './upload/upload.widget';
export { TransferWidget } from './transfer/transfer.widget';
export { SliderWidget } from './slider/slider.widget';
export { RateWidget } from './rate/rate.widget';
export { AutoCompleteWidget } from './autocomplete/autocomplete.widget';
export { CascaderWidget } from './cascader/cascader.widget';
export { MentionWidget } from './mention/mention.widget';
export { CustomWidget } from './custom/custom.widget';
export { NzWidgetRegistry } from './nz-widget.registry';
