// @delon/form/src/src/widgets/custom/sf-template.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, Input, TemplateRef } from '@angular/core';
import { SFComponent } from '../../sf.component';
var SFTemplateDirective = /** @class */ (function () {
    function SFTemplateDirective(templateRef, table) {
        this.templateRef = templateRef;
        this.table = table;
    }
    /**
     * @return {?}
     */
    SFTemplateDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.table._addTpl(this.path.startsWith('/') ? this.path : "/" + this.path, this.templateRef);
    };
    SFTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[sf-template]',
                },] },
    ];
    /** @nocollapse */
    SFTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
        { type: SFComponent, },
    ]; };
    SFTemplateDirective.propDecorators = {
        "path": [{ type: Input, args: ['sf-template',] },],
    };
    return SFTemplateDirective;
}());
export { SFTemplateDirective };
function SFTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SFTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SFTemplateDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SFTemplateDirective.propDecorators;
    /** @type {?} */
    SFTemplateDirective.prototype.path;
    /** @type {?} */
    SFTemplateDirective.prototype.templateRef;
    /** @type {?} */
    SFTemplateDirective.prototype.table;
}
