// @delon/abc/src/xlsx/xlsx.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DelonUtilModule } from '@delon/util';
import { XlsxService } from './xlsx.service';
import { XlsxDirective } from './xlsx.directive';
import { DA_XLSX_CONFIG } from './interface';
var /** @type {?} */ COMPONENTS = [XlsxDirective];
var AdXlsxModule = /** @class */ (function () {
    function AdXlsxModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AdXlsxModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: AdXlsxModule,
            providers: [XlsxService, { provide: DA_XLSX_CONFIG, useValue: config }],
        };
    };
    AdXlsxModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, DelonUtilModule],
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice(),
                },] },
    ];
    return AdXlsxModule;
}());
export { AdXlsxModule };
function AdXlsxModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AdXlsxModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AdXlsxModule.ctorParameters;
}
