// @delon/abc/src/tag-select/tag-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding, Input, Output, EventEmitter, } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
var TagSelectComponent = /** @class */ (function () {
    function TagSelectComponent() {
        this._expandable = true;
        this.expand = false;
        this.change = new EventEmitter();
    }
    Object.defineProperty(TagSelectComponent.prototype, "expandable", {
        get: /**
         * 是否启用 `展开与收进`
         * @return {?}
         */
        function () {
            return this._expandable;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._expandable = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TagSelectComponent.prototype.trigger = /**
     * @return {?}
     */
    function () {
        this.expand = !this.expand;
        this.change.emit(this.expand);
    };
    TagSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tag-select',
                    template: "\n    <ng-content></ng-content>\n    <a *ngIf=\"expandable\" class=\"trigger\" (click)=\"trigger()\">{{expand ? '\u6536\u8D77' : '\u5C55\u5F00'}}<i class=\"anticon anticon-{{expand ? 'up' : 'down'}} ml-sm\"></i></a>",
                    host: { '[class.ad-tag-select]': 'true' },
                    preserveWhitespaces: false,
                },] },
    ];
    /** @nocollapse */
    TagSelectComponent.propDecorators = {
        "expandable": [{ type: HostBinding, args: ['class.has-expand',] }, { type: Input },],
        "expand": [{ type: HostBinding, args: ['class.expanded',] },],
        "change": [{ type: Output },],
    };
    return TagSelectComponent;
}());
export { TagSelectComponent };
function TagSelectComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TagSelectComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TagSelectComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    TagSelectComponent.propDecorators;
    /** @type {?} */
    TagSelectComponent.prototype._expandable;
    /** @type {?} */
    TagSelectComponent.prototype.expand;
    /** @type {?} */
    TagSelectComponent.prototype.change;
}
