// @delon/abc/src/simple-table/interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * 数据源
 * @record
 */
export function SimpleTableData() { }
function SimpleTableData_tsickle_Closure_declarations() {
    /**
     * 选择框或单选框状态值
     * @type {?|undefined}
     */
    SimpleTableData.prototype.checked;
    /**
     * 选择框或单选框 `disabled` 值
     * @type {?|undefined}
     */
    SimpleTableData.prototype.disabled;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * 列描述
 * @record
 */
export function SimpleTableColumn() { }
function SimpleTableColumn_tsickle_Closure_declarations() {
    /**
     * 类型
     * - `checkbox` 多选
     * - `radio` 单选
     * - `link` 链接，可触发 `click`
     * - `img` 图片且居中(若 `className` 存在则优先)
     * - `number` 数字且居右(若 `className` 存在则优先)
     * - `currency` 货币且居右(若 `className` 存在则优先)
     * - `date` 日期格式且居中(若 `className` 存在则优先)
     * - `yn` 将`boolean`类型徽章化 [document](http://ng-alain.com/docs/data-render#yn)
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.type;
    /**
     * 列标题
     * @type {?}
     */
    SimpleTableColumn.prototype.title;
    /**
     * 列标题 i18n
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.i18n;
    /**
     * 列数据在数据项中对应的 key，支持 `a.b.c` 的嵌套写法
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.index;
    /**
     * 链接回调
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.click;
    /**
     * 按钮组
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.buttons;
    /**
     * 自定义渲染ID
     * \@example
     * ```html
     * <ng-template st-row="custom" let-item let-index="index" let-column="column">
     *  {{ c.title }}
     * </ng-template>
     * ```
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.render;
    /**
     * 标题自定义渲染ID
     * \@example
     * ```html
     * <ng-template st-row="custom" type="title" let-c>
     *  {{ item | json }}
     * </ng-template>
     * ```
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.renderTitle;
    /**
     * 列宽，例如：`10%`、`100px`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.width;
    /**
     * 排序的默认受控属性
     * - 只支持同时对一列进行排序，除非指定 `multiSort`，建议后端支持时使用
     * - 保证只有一列的 `sort` 值，否则自动获取所有列的第一个值
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.sort;
    /**
     * 排序函数，本地排序使用一个函数(参考 [Array.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) 的 compareFunction)
     * - 只有当属性存在时排序才会真的生效
     * - 如果是AJAX直接返回 true
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.sorter;
    /**
     * 排序的后端相对应的KEY，默认使用 `index` 属性
     * sortKey: 'name' => ?name=1&pi=1
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.sortKey;
    /**
     * 排序的后端相对应的VALUE
     * { ascend: '0', descend: '1' } => ?name=1&pi=1
     * { ascend: 'asc', descend: 'desc' } => ?name=desc&pi=1
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.sortReName;
    /**
     * 表头的筛选菜单项，至少一项以上才会生效
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filters;
    /**
     * 本地模式下，确定筛选的运行函数
     * - 只有当属性存在时筛选才会真的生效
     * - 如果是AJAX直接返回 true
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filter;
    /**
     * 标识数据是否经过过滤，筛选图标会高亮
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filtered;
    /**
     * 自定义 filter 图标，默认 `anticon anticon-filter`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filterIcon;
    /**
     * filter 确认按钮文本，默认 `确认`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filterConfirmText;
    /**
     * filter 清除按钮文本，默认 `重置`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filterClearText;
    /**
     * 是否多选，默认 `true`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filterMultiple;
    /**
     * 过滤的后端相对应的KEY，默认使用 `index` 属性
     * filterKey: 'name' => ?name=1&pi=1
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filterKey;
    /**
     * 过滤的后端相对应的VALUE
     * - 默认当 `filterMultiple` 时以英文逗号拼接的字符串
     * - 参数为 `filters` 原样数组
     * - 返回为 Object 对象
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.filterReName;
    /**
     * 格式化列值
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.format;
    /**
     * 自定义全/反选选择项
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.selections;
    /**
     * 列 `class` 属性值（注：无须 `.` 点），例如：
     * - `text-center` 居中
     * - `text-right` 居右
     * - `text-success` 成功色
     * - `text-danger` 异常色
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.className;
    /**
     * 合并列
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.colSpan;
    /**
     * 数字格式，`type=number` 有效
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.numberDigits;
    /**
     * 日期格式，`type=date` 有效，（默认：YYYY-MM-DD HH:mm）
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.dateFormat;
    /**
     * 真值条件，`type=yn` 有效，（默认：`true`）
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.ynTruth;
    /**
     * 徽章 `true` 时文本，`type=yn` 有效，（默认：是）
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.ynYes;
    /**
     * 徽章 `false` 时文本，`type=yn` 有效，（默认：否）
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.ynNo;
    /**
     * 是否允许导出，默认 `true`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.exported;
    /**
     * 权限，等同 `can()` 参数值
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.acl;
    /**
     * 当不存在数据时以默认值替代
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.default;
    /**
     * 固定前后列，当指定时务必指定 `width` 否则视为无效，有若干 **注意：** 项：
     *
     * - 若列头与内容不对齐或出现列重复，请指定列的宽度 `width`
     * - 建议指定 `scroll.x` 为大于表格宽度的固定值或百分比。注意，且非固定列宽度之和不要超过 `scroll.x`
     * @type {?|undefined}
     */
    SimpleTableColumn.prototype.fixed;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * 选择功能配置
 * @record
 */
export function SimpleTableSelection() { }
function SimpleTableSelection_tsickle_Closure_declarations() {
    /**
     * 选择项显示的文字
     * @type {?}
     */
    SimpleTableSelection.prototype.text;
    /**
     * 选择项点击回调
     * @type {?}
     */
    SimpleTableSelection.prototype.select;
    /**
     * 权限，等同 `can()` 参数值
     * @type {?|undefined}
     */
    SimpleTableSelection.prototype.acl;
}
/**
 * 过滤项配置
 * @record
 */
export function SimpleTableFilter() { }
function SimpleTableFilter_tsickle_Closure_declarations() {
    /**
     * 文本
     * @type {?}
     */
    SimpleTableFilter.prototype.text;
    /**
     * 值
     * @type {?|undefined}
     */
    SimpleTableFilter.prototype.value;
    /**
     * 是否选中
     * @type {?|undefined}
     */
    SimpleTableFilter.prototype.checked;
    /**
     * 权限，等同 `can()` 参数值
     * @type {?|undefined}
     */
    SimpleTableFilter.prototype.acl;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * 按钮配置
 * @record
 */
export function SimpleTableButton() { }
function SimpleTableButton_tsickle_Closure_declarations() {
    /**
     * 文本
     * @type {?}
     */
    SimpleTableButton.prototype.text;
    /**
     * 文本 i18n
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.i18n;
    /**
     * 格式化文本，较高调用频率，请勿过多复杂计算免得产生性能问题
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.format;
    /**
     * 按钮类型
     * - `none` 无任何互动
     * - `del` 删除，默认开启 `pop: true`
     * - `modal` 对话框，需要指定 `component` 才会生效
     * - `static` 静态对话框，需要指定 `component` 才会生效
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.type;
    /**
     * 点击回调
     * - Function
     *  - `type=modal` 只会在当有传回值时才会触发回调
     * - reload：重新刷新当前页
     * - load：重新加载数据，并重置页码为：`1`
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.click;
    /**
     * 是否需要气泡确认框
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.pop;
    /**
     * 气泡确认框内容，默认 `确认删除吗？`
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.popTitle;
    /**
     * 对话框组件对象，务必在 `entryComponents` 注册
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.component;
    /**
     * 对话框参数
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.params;
    /**
     * 指定模态框目标组件的接收参数名，默认：`record`
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.paramName;
    /**
     * 对话框大小，默认：`lg`
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.size;
    /**
     * 对话框额外参数，见 [ModalHelper](http://ng-alain.com/docs/service#ModalHelper)
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.modalOptions;
    /**
     * 下拉菜单，当存在时以 `dropdown` 形式渲染
     * - 只支持一级
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.children;
    /**
     * 权限，等同 `can()` 参数值
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.acl;
    /**
     * 条件表达式，较高调用频率，请勿过多复杂计算免得产生性能问题
     * @type {?|undefined}
     */
    SimpleTableButton.prototype.if;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * 回调数据
 * @record
 */
export function SimpleTableChange() { }
function SimpleTableChange_tsickle_Closure_declarations() {
    /**
     * 回调类型
     * @type {?}
     */
    SimpleTableChange.prototype.type;
    /**
     * 当前页码
     * @type {?}
     */
    SimpleTableChange.prototype.pi;
    /**
     * 每页数量
     * @type {?}
     */
    SimpleTableChange.prototype.ps;
    /**
     * 数据总量
     * @type {?}
     */
    SimpleTableChange.prototype.total;
}
/**
 * @record
 */
export function ReqReNameType() { }
function ReqReNameType_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ReqReNameType.prototype.pi;
    /** @type {?|undefined} */
    ReqReNameType.prototype.ps;
}
/**
 * @record
 */
export function ResReNameType() { }
function ResReNameType_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ResReNameType.prototype.total;
    /** @type {?|undefined} */
    ResReNameType.prototype.list;
}
/**
 * @record
 */
export function STExportOptions() { }
function STExportOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    STExportOptions.prototype._d;
    /** @type {?} */
    STExportOptions.prototype._c;
    /**
     * 工作溥名
     * @type {?|undefined}
     */
    STExportOptions.prototype.sheetname;
    /**
     * 文件名
     * @type {?|undefined}
     */
    STExportOptions.prototype.filename;
    /**
     * triggers when saveas
     * @type {?|undefined}
     */
    STExportOptions.prototype.callback;
}
