// @delon/abc/src/simple-table/simple-table.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DelonUtilModule } from '@delon/util';
import { SimpleTableComponent } from './simple-table.component';
import { SimpleTableRowDirective } from './simple-table-row.directive';
import { AdSimpleTableConfig } from './simple-table.config';
var /** @type {?} */ COMPONENTS = [SimpleTableComponent, SimpleTableRowDirective];
import { NgZorroAntdModule } from 'ng-zorro-antd';
var /** @type {?} */ ZORROMODULES = [NgZorroAntdModule];
var AdSimpleTableModule = /** @class */ (function () {
    function AdSimpleTableModule() {
    }
    /**
     * @return {?}
     */
    AdSimpleTableModule.forRoot = /**
     * @return {?}
     */
    function () {
        return { ngModule: AdSimpleTableModule, providers: [AdSimpleTableConfig] };
    };
    AdSimpleTableModule.decorators = [
        { type: NgModule, args: [{
                    schemas: [NO_ERRORS_SCHEMA],
                    imports: [CommonModule, FormsModule, DelonUtilModule].concat(ZORROMODULES),
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice(),
                },] },
    ];
    return AdSimpleTableModule;
}());
export { AdSimpleTableModule };
function AdSimpleTableModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AdSimpleTableModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AdSimpleTableModule.ctorParameters;
}
