// @delon/abc/src/simple-table/simple-table-row.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, Input, TemplateRef } from '@angular/core';
import { SimpleTableComponent } from './simple-table.component';
var SimpleTableRowDirective = /** @class */ (function () {
    function SimpleTableRowDirective(templateRef, table) {
        this.templateRef = templateRef;
        this.table = table;
    }
    /**
     * @return {?}
     */
    SimpleTableRowDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ col = this.table._columns.find(function (w) { return w.render === _this.id; });
        if (!col)
            return;
        switch (this.type) {
            case 'title':
                col["__renderTitle"] = this.templateRef;
                break;
            default:
                col["__render"] = this.templateRef;
                break;
        }
    };
    SimpleTableRowDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[st-row]',
                },] },
    ];
    /** @nocollapse */
    SimpleTableRowDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
        { type: SimpleTableComponent, },
    ]; };
    SimpleTableRowDirective.propDecorators = {
        "id": [{ type: Input, args: ['st-row',] },],
        "type": [{ type: Input },],
    };
    return SimpleTableRowDirective;
}());
export { SimpleTableRowDirective };
function SimpleTableRowDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SimpleTableRowDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SimpleTableRowDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SimpleTableRowDirective.propDecorators;
    /** @type {?} */
    SimpleTableRowDirective.prototype.id;
    /** @type {?} */
    SimpleTableRowDirective.prototype.type;
    /** @type {?} */
    SimpleTableRowDirective.prototype.templateRef;
    /** @type {?} */
    SimpleTableRowDirective.prototype.table;
}
