// @delon/acl/src/acl.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DelonACLConfig } from './acl.config';
import { ACLGuard } from './services/acl-guard';
import { ACLService } from './services/acl.service';
var /** @type {?} */ SERVICES = [ACLService, ACLGuard];
import { ACLDirective } from './directives/acl.directive';
var /** @type {?} */ COMPONENTS = [ACLDirective];
var DelonACLModule = /** @class */ (function () {
    function DelonACLModule() {
    }
    /**
     * @return {?}
     */
    DelonACLModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: DelonACLModule,
            providers: [DelonACLConfig].concat(SERVICES),
        };
    };
    DelonACLModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice(),
                },] },
    ];
    return DelonACLModule;
}());
export { DelonACLModule };
function DelonACLModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DelonACLModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DelonACLModule.ctorParameters;
}
